File: source/includes/read/change-streams.c
#include <stdio.h>
#include <bson/bson.h>
#include <mongoc/mongoc.h>

int
main (void)
{
    mongoc_client_t *client;
    mongoc_collection_t *collection;
    mongoc_init ();

    client = mongoc_client_new ("<connection string URI>");
    collection = mongoc_client_get_collection (client, "sample_restaurants", "restaurants");

    {
        // Opens a change stream on the collection and prints each change
        // start-open-change-stream
        bson_t *pipeline = bson_new ();
        const bson_t *doc;

        mongoc_change_stream_t *change_stream =
            mongoc_collection_watch (collection, pipeline, NULL);

        while (true) {
            bson_error_t error;
            if (mongoc_change_stream_next (change_stream, &doc)) {
                char *str = bson_as_canonical_extended_json (doc, NULL);
                printf ("Received change: %s\n", str);
                bson_free (str);
            } else if (mongoc_change_stream_error_document (change_stream, &error, NULL)) {
                printf("Got error on change stream: %s\n", error.message);
                break;
            }
        }

        bson_destroy (pipeline);
        mongoc_change_stream_destroy (change_stream);
        // end-open-change-stream
    }
    {
        // Updates a document in the collection to trigger a change event
        // start-update-for-change-stream
        bson_t *filter = BCON_NEW ("name", BCON_UTF8 ("Blarney Castle"));
        bson_t *update = BCON_NEW("$set", "{", "cuisine", BCON_UTF8 ("Irish"), "}");

        mongoc_collection_update_one (collection, filter, update, NULL, NULL, NULL);
        // end-update-for-change-stream
    }
    {
        // Opens a change stream on the collection and specifies a pipeline to only receive update events
        // start-change-stream-pipeline
        bson_t *pipeline = BCON_NEW (
            "pipeline", "[", 
            "{", "$match", "{", "operationType", BCON_UTF8 ("update"), "}", "}", 
            "]");
        const bson_t *doc;

        mongoc_change_stream_t *change_stream =
            mongoc_collection_watch (collection, pipeline, NULL);

        while (mongoc_change_stream_next (change_stream, &doc)) {
          char *str = bson_as_canonical_extended_json (doc, NULL);
          printf ("Received change: %s\n", str);
          bson_free (str);
        }

        bson_destroy (pipeline);
        mongoc_change_stream_destroy (change_stream);
        // end-change-stream-pipeline
    }
    {
        // Opens a change stream on the collection and specifies options to receive the full document for update events
        // start-change-stream-post-image
        bson_t *pipeline = bson_new ();
        bson_t *opts = BCON_NEW ("fullDocument", BCON_UTF8 ("updateLookup"));
        const bson_t *doc;

        mongoc_change_stream_t *change_stream =
            mongoc_collection_watch (collection, pipeline, opts);

        while (true) {
            bson_error_t error;
            if (mongoc_change_stream_next (change_stream, &doc)) {
                char *str = bson_as_canonical_extended_json (doc, NULL);
                printf ("Received change: %s\n", str);
                bson_free (str);
            } else if (mongoc_change_stream_error_document (change_stream, &error, NULL)) {
                printf("Got error on change stream: %s\n", error.message);
                break;
            }
        }

        bson_destroy (pipeline);
        bson_destroy (opts);
        mongoc_change_stream_destroy (change_stream);
        // end-change-stream-post-image
    }

    mongoc_collection_destroy (collection);
    mongoc_client_destroy (client);
    mongoc_cleanup ();

    return EXIT_SUCCESS;
}

File: v3.3/source/includes/code-examples/change-streams/change-streams.cs
// start-open-change-stream
var database = client.GetDatabase("sample_restaurants");
var collection = database.GetCollection<Restaurant>("restaurants");

// Opens a change stream and prints the changes as they're received
using (var cursor = collection.Watch())
{
    foreach (var change in cursor.ToEnumerable())
    {
        Console.WriteLine("Received the following type of change: " + change.BackingDocument);
    }
}
// end-open-change-stream

// start-open-change-stream-async
var database = client.GetDatabase("sample_restaurants");
var collection = database.GetCollection<Restaurant>("restaurants");

// Opens a change streams and print the changes as they're received
using var cursor = await collection.WatchAsync();
await cursor.ForEachAsync(change =>
{
    Console.WriteLine("Received the following type of change: " + change.BackingDocument);
});
// end-open-change-stream-async

// start-modify-document
var database = client.GetDatabase("sample_restaurants");
var collection = database.GetCollection<Restaurant>("restaurants");

var filter = Builders<Restaurant>.Filter
    .Eq(restaurant => restaurant.Name, "Blarney Castle");

var update = Builders<Restaurant>.Update
    .Set(restaurant => restaurant.Cuisine, "Irish");

var result = collection.UpdateOne(filter, update);
// end-modify-document

// start-change-stream-pipeline-async
var pipeline = new EmptyPipelineDefinition<ChangeStreamDocument<Restaurant>>()
            .Match(change => change.OperationType == ChangeStreamOperationType.Update);

// Opens a change stream and prints the changes as they're received
using (var cursor = await collection.WatchAsync(pipeline))
{
    await cursor.ForEachAsync(change =>
    {
        Console.WriteLine("Received the following change: " + change);
    });
}
// end-change-stream-pipeline-async

// start-change-stream-pipeline
var pipeline = new EmptyPipelineDefinition<ChangeStreamDocument<Restaurant>>()
            .Match(change => change.OperationType == ChangeStreamOperationType.Update);

// Opens a change streams and print the changes as they're received
using (var cursor = collection.Watch(pipeline))
{
    foreach (var change in cursor.ToEnumerable())
    {
        Console.WriteLine("Received the following change: " + change);
    }
}
// end-change-stream-pipeline

// start-split-event-helpers-sync
// Fetches the next complete change stream event
private static IEnumerable<ChangeStreamDocument<TDocument>> GetNextChangeStreamEvent<TDocument>(
IEnumerator<ChangeStreamDocument<TDocument>> changeStreamEnumerator)
{
    while (changeStreamEnumerator.MoveNext())
    {
        var changeStreamEvent = changeStreamEnumerator.Current;
        if (changeStreamEvent.SplitEvent != null)
        {
            var fragment = changeStreamEvent;
            while (fragment.SplitEvent.Fragment < fragment.SplitEvent.Of)
            {
                changeStreamEnumerator.MoveNext();
                fragment = changeStreamEnumerator.Current;
                MergeFragment(changeStreamEvent, fragment);
            }
        }
        yield return changeStreamEvent;
    }
}

// Merges a fragment into the base event
private static void MergeFragment<TDocument>(
    ChangeStreamDocument<TDocument> changeStreamEvent,
    ChangeStreamDocument<TDocument> fragment)
{
    foreach (var element in fragment.BackingDocument)
    {
        if (element.Name != "_id" && element.Name != "splitEvent")
        {
            changeStreamEvent.BackingDocument[element.Name] = element.Value; 
        }
    }
}
// end-split-event-helpers-sync

// start-split-event-helpers-async
// Fetches the next complete change stream event
private static async IAsyncEnumerable<ChangeStreamDocument<TDocument>> GetNextChangeStreamEventAsync<TDocument>(
    IAsyncCursor<ChangeStreamDocument<TDocument>> changeStreamCursor)
{
    var changeStreamEnumerator = GetNextChangeStreamEventFragmentAsync(changeStreamCursor).GetAsyncEnumerator();
    while (await changeStreamEnumerator.MoveNextAsync())
    {
        var changeStreamEvent = changeStreamEnumerator.Current;
        if (changeStreamEvent.SplitEvent != null)
        {
            var fragment = changeStreamEvent;
            while (fragment.SplitEvent.Fragment < fragment.SplitEvent.Of)
            {
                await changeStreamEnumerator.MoveNextAsync();
                fragment = changeStreamEnumerator.Current;
                MergeFragment(changeStreamEvent, fragment);
            }
        }
        yield return changeStreamEvent;
    }
}

private static async IAsyncEnumerable<ChangeStreamDocument<TDocument>> GetNextChangeStreamEventFragmentAsync<TDocument>(
    IAsyncCursor<ChangeStreamDocument<TDocument>> changeStreamCursor)
{
    while (await changeStreamCursor.MoveNextAsync())
    {
        foreach (var changeStreamEvent in changeStreamCursor.Current)
        {
            yield return changeStreamEvent;
        }
    }
}

// Merges a fragment into the base event
private static void MergeFragment<TDocument>(
    ChangeStreamDocument<TDocument> changeStreamEvent,
    ChangeStreamDocument<TDocument> fragment)
{
    foreach (var element in fragment.BackingDocument)
    {
        if (element.Name != "_id" && element.Name != "splitEvent")
        {
            changeStreamEvent.BackingDocument[element.Name] = element.Value;
        }
    }
}
// end-split-event-helpers-async

// start-split-change-event-sync
var pipeline = new EmptyPipelineDefinition<ChangeStreamDocument<Restaurant>>()
    .ChangeStreamSplitLargeEvent();

using var cursor = collection.Watch(pipeline);
foreach (var completeEvent in GetNextChangeStreamEvent(cursor.ToEnumerable().GetEnumerator()))
{
    Console.WriteLine("Received the following change: " + completeEvent.BackingDocument);
}
// end-split-change-event-sync

// start-split-change-event-async
var pipeline = new EmptyPipelineDefinition<ChangeStreamDocument<Restaurant>>()
    .ChangeStreamSplitLargeEvent();

using var cursor = await collection.WatchAsync(pipeline);
await foreach (var completeEvent in GetNextChangeStreamEventAsync(cursor))
{
    Console.WriteLine("Received the following change: " + completeEvent.BackingDocument);
}
// end-split-change-event-async

// start-change-stream-post-image
var pipeline = new EmptyPipelineDefinition<ChangeStreamDocument<Restaurant>>()
    .Match(change => change.OperationType == ChangeStreamOperationType.Update);

var options = new ChangeStreamOptions
{
    FullDocument = ChangeStreamFullDocumentOption.UpdateLookup,
};

using (var cursor = collection.Watch(pipeline, options))
{
    foreach (var change in cursor.ToEnumerable())
    {
        Console.WriteLine(change.FullDocument.ToBsonDocument());
    }
}
// end-change-stream-post-image

// start-change-stream-post-image-async
var pipeline = new EmptyPipelineDefinition<ChangeStreamDocument<Restaurant>>()
    .Match(change => change.OperationType == ChangeStreamOperationType.Update);

var options = new ChangeStreamOptions
{
    FullDocument = ChangeStreamFullDocumentOption.UpdateLookup,
};

using var cursor = await collection.WatchAsync(pipeline, options);
await cursor.ForEachAsync(change =>
{
    Console.WriteLine(change.FullDocument.ToBsonDocument());
});
// end-change-stream-post-image-async